Add support for exporting compiled examples

brentru opened the pull request:
This pull request adds support for exporting compiled examples from CI-Arduino.

Compilation:

- If a `.platform.generate` file is included in the `libraryName/examples` folder, Arduino-CLI will export the compiled example as a `.hex` file. 
- Platforms that support the UF2 file format will additionally pack the .hex into a .uf2 file.
- The arduino-cli version has been updated to v18.3

----

Distribution:

- Repositories that include an upload-artifact step (see below) will attach the `.hex` and `.uf2` files to the build in a .zip file (Example, under "Artifacts": https://github.com/brentru/Adafruit_SI1145_Library/actions/runs/1222715452)

For example, the repository above contains two generation files: `.uno.generate` and `.metro_m4.generate`. The AVR (uno) only produces a .hex file:
![hex_uno_export](https://user-images.githubusercontent.com/322428/132879023-629205f4-bff6-4df9-b611-bbf8742e64a0.png)


And the Metro M4 (SAMD51) produces a .hex and .uf2 file:
![metro_export](https://user-images.githubusercontent.com/322428/132878992-8d801f21-f59c-42c9-8547-4fac9c1afba6.png)

----

To attach the compiled build artifacts to a run, use the [upload-artifact](https://github.com/actions/upload-artifact) GitHub action:
```
    - name: Upload build artifacts
      uses: actions/upload-artifact@v2
      with:
        name: ${{ github.event.repository.name }}.${{ github.sha }}
        path: |
            examples/*/build/*/*.hex
            examples/*/build/*/*.bin
            examples/*/build/*/*.uf2
```

ladyada reviewed (approved):
why not, lets give it a shot! :)

brentru commented:
* Added QTPY (SAMD21)
* Fixed previous issue with ESP32 binary generation
* Updated YML in instructions and example Adafruit_SI1145_Library to reflect artifact naming scheme: `REPO_NAME.SHA`
